Inline case search for child modules

Jtang-1 opened the pull request:
## Product Description
<!-- For non-invisible changes, describe user-facing effects. -->
Adds compatibility using sub-modules with parent menu configured with inline case search.

Adds app builder input for custom instance name that will be applied to the `results` and `search-input:results` instances. 

<img width="706" alt="Screen Shot 2023-10-20 at 3 21 47 PM" src="https://github.com/dimagi/commcare-hq/assets/88759246/cc14b029-e61f-4697-88fb-8f14078c6b4b">
 

## Technical Summary
<!--
    Provide a link to the ticket or document which prompted this change,
    Describe the rationale and design decisions.
-->
[Tech Spec](https://docs.google.com/document/d/1hp6b3ilLFDF6YYwLLgcr18v4SNR089lOBlLrYDZMOBw/edit)

Jira Ticket [USH-3712](https://dimagi-dev.atlassian.net/browse/USH-3712)
## Feature Flag
<!-- If this is specific to a feature flag, which one? -->
[Inline Case Search](https://www.commcarehq.org/hq/flags/edit/inline_case_search/)

## Safety Assurance

### Safety story
<!--
Describe how you became confident in this change, such as
local testing, why the change is inherently safe, and/or plans to limit the blast radius of a defect.

In particular consider how existing data may be impacted by this change.
-->
The blast radius is limited to the inline case search FF.

### Automated test coverage

<!-- Identify the related test coverage and the tests it would catch -->
Test added for conflicting instance name app build validation.
### QA Plan

<!--
- Describe QA plan that along with automated test coverages proves this PR is regression free
- Link to QA Ticket
-->
No QA.
### Rollback instructions

<!--
If this PR follows standards of revertability, check the box below.
Otherwise replace it with detailed instructions or reasons a rollback is impossible.
-->

- [X] This PR can be reverted after deploy with no further considerations

### Labels & Review
- [X] Risk label is set correctly
- [X] The set of people pinged as reviewers is appropriate for the level of risk of the change


[USH-3712]: https://dimagi-dev.atlassian.net/browse/USH-3712?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

esoergel committed:
add instance_name to model

esoergel committed:
Add basic UI for customizing instance names

I'd like to revisit the help text once the rest of this work is done -
maybe show an example of what the actual instance ID will be, maybe a
placeholder, etc

esoergel committed:
lint

Jtang-1 committed:
feat: uses user-defined results instance name as query block storage-instance

Jtang-1 committed:
feat: renames session datum nodeset to match query storage-instance

Jtang-1 committed:
feat: uses user defined instance name for detail block

Jtang-1 committed:
feat: build validation for allowable instance-names for result and search-input instances

Jtang-1 committed:
refactor: pulls out adding post to entry

Jtang-1 committed:
refactor: move pulled results instance name validation out as a utility function

Jtang-1 committed:
bug: moves import to avoid circular import

Jtang-1 committed:
feat: adds build validation if submodule has same instance name as root module

Jtang-1 committed:
test: updates test for conflicting root and submodule instance name

Jtang-1 committed:
lint

snopoke reviewed (commented):
Some suite xml tests would be great.

Jtang-1 committed:
bug: uses custom instance name to entry post block if relevant

Jtang-1 committed:
refactor: moves getting instance name logic to CaseSearch

Jtang-1 committed:
test: adds test that custom instance is added to suite

Jtang-1 commented on corehq/apps/app_manager/suite_xml/sections/entries.py:
```diff
@@ -241,6 +242,19 @@ def include_post_in_entry(self, module_id):
+        )
+        case_session_var = self.get_case_session_var_for_form(form)
+        storage_instance = RESULTS_INSTANCE_INLINE if module_uses_inline_search(module) \
+            else 'casedb'
```
yes, thanks for catching that [c89b907](https://github.com/dimagi/commcare-hq/pull/33635/commits/c89b9074129295b84fec5efdc4981a73366c4040)

Jtang-1 commented on corehq/apps/app_manager/models.py:
```diff
@@ -2020,7 +2021,10 @@ def get_instance_name(self, module):
+                if module.search_config.instance_name:
+                    return f'{RESULTS_INSTANCE_BASE}{module.search_config.instance_name}'
+                else:
+                    return RESULTS_INSTANCE_INLINE
```
[bfbf126](https://github.com/dimagi/commcare-hq/pull/33635/commits/bfbf12698f53958dbbc137c21dc9bb197b8d6143)

esoergel committed:
Check that instance name is valid

Jtang-1 committed:
bug: checks search_config attribute exists before accessing its properties

Jtang-1 committed:
bug: assigns query_datum_id before using

esoergel commented on corehq/apps/app_manager/suite_xml/post_process/instances.py:
```diff
@@ -354,9 +356,13 @@ def search_input_instances(app, instance_name):
+        is_valid_instance_name = True
+    if is_valid_instance_name:
+        return Instance(id=instance_name, src=src)
+    return None
```
This function reads a little weird - there are two fallback options, the legacy instance and `None`, and it's a bit difficult to tell which should be returned when.  Would it be fair to write this as:

```python
def search_input_instances(app, instance_name):
    if ":" not in instance_name:
        return Instance(id=instance_name, src='jr://instance/search-input')

    _, query_datum_id = instance_name.split(':', 1)
    if is_valid_results_instance_name(app, query_datum_id):
        src = f'jr://instance/search-input/{query_datum_id}'
        return Instance(id=instance_name, src=src)
    return None
```

Jtang-1 replied:
yes, [c5dba28](https://github.com/dimagi/commcare-hq/pull/33635/commits/c5dba28dbe1039bdec6a6e0cf9e1ec89bb255bbc)

esoergel commented on corehq/apps/app_manager/suite_xml/post_process/instances.py:
```diff
@@ -366,7 +372,9 @@ def selected_cases_instances(app, instance_name):
-    return Instance(id=instance_name, src=f'jr://instance/remote/{instance_name}')
+    if is_valid_results_instance_name(app, instance_name):
+        return Instance(id=instance_name, src=f'jr://instance/remote/{instance_name}')
+    return None
```
Haven't read over the whole PR yet, so this might be in there somewhere, but as a reminder to myself, what happens when this returns `None`?  Should it instead raise some exception (whether a 400 or 500)?

Jtang-1 replied:
None is returned [here](https://github.com/dimagi/commcare-hq/blob/3771ff8f14c4ae6d3db8f2451cc83c634c956f45/corehq/apps/app_manager/suite_xml/post_process/instances.py#L425) and the instance name is returned in `unknown_instance_ids` when it's called by[add_missing_instances](https://github.com/dimagi/commcare-hq/blob/b719aeaac9243114e56d0ea4d02106ed1ca08555/corehq/apps/app_manager/xform.py#L869). This ends up raising [XFormValidationError](https://github.com/dimagi/commcare-hq/blob/b719aeaac9243114e56d0ea4d02106ed1ca08555/corehq/apps/app_manager/xform.py#L879)

Jtang-1 replied:
I considered raising the `UnknownInstanceError` but decided to handle it the same way as `null factory` https://github.com/dimagi/commcare-hq/blob/3771ff8f14c4ae6d3db8f2451cc83c634c956f45/corehq/apps/app_manager/suite_xml/post_process/instances.py#L317

esoergel replied:
cool thanks - I do now see that the reports instance factory similarly returns `None` in some cases.

Jtang-1 committed:
refactor: improves logic flow

esoergel committed:
Add example reference and more help text

Jtang-1 committed:
test: adds second module with unique instance_name and pulls out test

esoergel commented on corehq/apps/app_manager/suite_xml/sections/entries.py:
```diff
@@ -545,6 +550,18 @@ def add_remote_query_datums(self, datums):
+            return datum
+        instance_pattern = r"""instance\(['"]results:(.*?)['"]\)"""
+        new_instance = f"instance('{instance_name}')"
+        renamed = re.sub(instance_pattern, new_instance, datum.datum.nodeset)
```
know where these `datum`s are coming from?  Is it possible to instead have them use the correct instance name up front rather than modify that later?  Similarly, is there any risk associated with mutating the datums?  If they're accessed elsewhere, the instance name used will depend on whether that's called before or after this.

Jtang-1 replied:
the `datum` are created via [get_case_datums_basic_module](https://github.com/dimagi/commcare-hq/blob/d35070e90dbcad8d231be216e0d913e9279b3e84/corehq/apps/app_manager/suite_xml/sections/entries.py#L518) which generates/returns all the datum within the <session> block (for that module/form) During the datum creation, it seems we do know whether the module is using inline_case_search so I was able to name it upfront [d35070e](https://github.com/dimagi/commcare-hq/pull/33635/commits/d35070e90dbcad8d231be216e0d913e9279b3e84)

esoergel commented on corehq/apps/app_manager/suite_xml/sections/entries.py:
```diff
@@ -656,10 +673,9 @@ def get_query_datums(self, module, uses_inline_search):
             RemoteRequestFactory,
         )
-        storage_instance = RESULTS_INSTANCE_INLINE if uses_inline_search else RESULTS_INSTANCE
+        storage_instance = module.search_config.get_instance_name() if uses_inline_search else RESULTS_INSTANCE
```
Would it be advisable to move this `if/else` logic into `get_instance_name()`, either by passing in `uses_inline_search` or looking it up in the body of that method.  Looks like this logic is applied in each of the references to it, though maybe with subtle differences that matter, I'm not certain

Jtang-1 replied:
I considered it, but the `else` value isn't always `RESULTS_INSTANCE`, and there's a unique logic for each of the callers to decide what value to assign.
For example in this  [usage](https://github.com/dimagi/commcare-hq/blob/bfbf12698f53958dbbc137c21dc9bb197b8d6143/corehq/apps/app_manager/suite_xml/sections/entries.py#L250), that `else` value would need to be passed into `get_instance_name()` or a flag to know to use `casedb` instead of `RESULTS_INSTANCE` which I don't think is readable.

esoergel commented on corehq/apps/app_manager/suite_xml/utils.py:
```diff
@@ -119,3 +119,17 @@ def get_ordered_case_types(case_type, additional_case_types=None):
+    )
+
+    valid_instance_names = {RESULTS_INSTANCE, RESULTS_INSTANCE_INLINE}
+    valid_instance_names.update(f'{RESULTS_INSTANCE_BASE}{module.search_config.instance_name}'
```
should this be `get_instance_name()`? If not defined, this will add `results:None` (or possibly `results:` as valid names.

Jtang-1 replied:
Yes thanks for catching that, [e14c0d8](https://github.com/dimagi/commcare-hq/pull/33635/commits/e14c0d8c254ed6b5f94cdf3e6435e376e090b058)

Jtang-1 committed:
bug: uses get_instance_name() otherwise results:None would be considered valid

Jtang-1 committed:
refactor: consistency in getting instance name from search_config

Jtang-1 committed:
bug: get_instance_name() already includes base 'results:' in return

esoergel commented on corehq/apps/app_manager/suite_xml/utils.py:
```diff
@@ -126,10 +126,9 @@ def is_valid_results_instance_name(app, instance_name):
 
     valid_instance_names = {RESULTS_INSTANCE, RESULTS_INSTANCE_INLINE}
-    valid_instance_names.update(f'{RESULTS_INSTANCE_BASE}{module.search_config.get_instance_name()}'
+    valid_instance_names.update(f'{module.search_config.get_instance_name()}'
```
nit: you can also drop the `f'{}'` from these now

Jtang-1 replied:
[4920797](https://github.com/dimagi/commcare-hq/pull/33635/commits/4920797c97538e25c9ae711ef5dc9b6cc93cd3f9)

Jtang-1 committed:
defines datum nodeset with instance name if module uses inline search.

The instance name used for the nodeset needs to be the same as the storage-instance.
The query datum is added into the entry block later via get_query_datums

Jtang-1 committed:
style: remove unnecessary f-string

esoergel commented on corehq/apps/app_manager/suite_xml/sections/entries.py:
```diff
@@ -628,7 +613,8 @@ def get_datum_meta_module(self, module, use_filter=False):
             if loads_registry_case or uses_inline_search:
                 if uses_inline_search:
-                    instance_name, root_element = "results:inline", "results"
+                    instance_name = detail_module.search_config.get_instance_name()
```
:star:
